File: SWEA/Java/[D3] 2817 부분 수열의 합 (재귀)/부분수열의합.java
import java.util.Scanner;

public class 부분수열의합 {
	// 자연수 개수
	static int N;

	// 부분 수열의 합
	static int K;

	// N개의 자연수
	static int[] nums;

	// 결과: 부분 수열의 합이 K가 되는 경우의 수
	static int cnt;

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);

		// 테스트케이스 개수
		int T = sc.nextInt();

		// T개의 테스트케이스
		for (int t = 1; t <= T; t++) {
			// 자연수 개수
			N = sc.nextInt();

			// 부분 수열의 합
			K = sc.nextInt();

			// N개의 자연수
			nums = new int[N];
			for (int i = 0; i < N; i++) {
				nums[i] = sc.nextInt();
			}

			// 결과: 부분 수열의 합이 K가 되는 경우의 수
			cnt = 0;
			powerSet(0, 0);

			System.out.printf("#%d %d\n", t, cnt);
		}
	}

	// 부분 수열의 합이 K가 되는 경우의 수
	static void powerSet(int idx, int sum) {
		// sum이 K가 되면 cnt 증가
		if (sum == K) {
			cnt++;
			return;
		}

		// sum이 K보다 크면 그만두기
		else if (sum > K) {
			return;
		}

		// idx가 인데스 범위를 넘어가면 그만두기
		if (idx >= N) {
			return;
		}

		// 현재 수 더하고 다음 인덱스로 넘어가기
		powerSet(idx + 1, sum + nums[idx]);

		// 현재 수 건너뛰기
		powerSet(idx + 1, sum);
	}
}

File: 백준/Python/[S2] 1182 부분수열의 합/부분수열의 합.py
import sys
input = sys.stdin.readline

# 정수 하나씩 선택하기
# idx: 선택하는 수의 인덱스
def get_sum(idx):
    global N, S, cnt

    # 모든 수를 탐색한 경우
    if idx == N:
        # 부분 수열의 합이 S인 경우
        if nums and sum(nums) == S:
            cnt += 1
        return

    # 현재 수를 선택하지 않는 경우
    get_sum(idx + 1)

    # 현재 수를 선택하는 경우
    nums.append(A[idx])
    get_sum(idx + 1)
    nums.pop()

##############################################

# 정수 개수, 수열의 합
N, S = map(int, input().split())

# N개의 정수
A = list(map(int, input().split()))

# 부분 수열
nums = []

# 합이 S가 되는 부분 수열의 개수
cnt = 0

# 정수 하나씩 선택하기
get_sum(0)

print(cnt)